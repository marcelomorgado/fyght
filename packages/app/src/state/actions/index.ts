export * from "./myFyghter";
export * from "./enemies";
export * from "./messages";
export * from "./metamask";
export * from "./daiBalances";
