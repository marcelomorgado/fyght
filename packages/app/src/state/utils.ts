/* eslint-disable no-undef */
import { TransactionReceipt } from "ethers/providers";
import { Transaction } from "ethers/utils";

const ETHEREUM_NETWORK_ID = process.env.ETHEREUM_NETWORK_ID;

// TODO: onFinally or just finally function
export const optimisticUpdate = async ({
  doTransaction,
  onOptimistic,
  onSuccess,
  onError,
  getState,
}: {
  doTransaction: () => Promise<Transaction>;
  onOptimistic?: () => void;
  onSuccess?: (receipt?: TransactionReceipt) => void;
  onError: (errorMessage: string, receipt?: TransactionReceipt) => void;
  getState: () => FyghtState;
}): Promise<void> => {
  const {
    metamask: { loomProvider, ethereumProvider },
  } = getState();

  if (onOptimistic) {
    onOptimistic();
  }

  try {
    const tx = await doTransaction();
    const { chainId } = tx;

    console.log(tx);

    const provider = `${chainId}` === ETHEREUM_NETWORK_ID ? ethereumProvider : loomProvider;

    provider.once(tx.hash, (receipt: TransactionReceipt) => {
      const { status } = receipt;
      if (!status) {
        onError("", receipt);
        return;
      }
      if (onSuccess) {
        onSuccess(receipt);
      }
    });
  } catch (e) {
    // Note: Keeping for now due debug purposes
    console.log(e);

    const errorMessage =
      e.data && e.data.message
        ? e.data.message.replace("VM Exception while processing transaction: revert ", "")
        : "Unexpected error";

    onError(errorMessage);
  }
};
