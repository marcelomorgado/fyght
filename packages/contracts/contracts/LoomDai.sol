pragma solidity 0.6.2;

import "@openzeppelin/contracts/math/SafeMath.sol";
import "@openzeppelin/contracts/token/ERC20/ERC20.sol";


contract LoomDai is ERC20 {
    // Transfer Gateway contract address
    address public gateway;

    constructor(address _gateway) public {
        gateway = _gateway;
    }

    // Used by the DAppChain Gateway to mint tokens that have been deposited to the Ethereum Gateway
    function mintToGateway(uint256 _amount) public {
        require(msg.sender == gateway, "only the gateway is allowed to mint");
        _mint(gateway, _amount);
    }
}
