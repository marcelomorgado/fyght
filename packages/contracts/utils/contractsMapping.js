const {
  Client,
  NonceTxMiddleware,
  SignedTxMiddleware,
  Address,
  LocalAddress,
  CryptoUtils,
  LoomProvider,
  Contracts,
  soliditySha3,
} = require("loom-js");
const Web3 = require("web3");
const dotenv = require("dotenv");
const { OfflineWeb3Signer } = require("loom-js/dist/solidity-helpers");

dotenv.config();

const {
  LOOM_NETWORK,
  ETHEREUM_NETWORK_ID,
  LOOM_NETWORK_ID,
  LOOM_WRITE_URL,
  LOOM_READ_URL,
  ETHEREUM_PRIVATE_KEY,
  LOOM_PRIVATE_KEY,
  INFURA_PROVIDER_URL,
} = process.env;

const { TransferGateway } = Contracts;

const EthereumDai = require("../build/contracts/EthereumDai.json");
const LoomDai = require("../build/contracts/LoomDai.json");

const {
  networks: {
    [ETHEREUM_NETWORK_ID]: { address: ethereumDaiAddress, transactionHash: ethereumDaiDeployTxHash },
  },
} = EthereumDai;

const {
  networks: {
    [LOOM_NETWORK_ID]: { address: loomDaiAddress },
  },
} = LoomDai;

function loadExtdevAccount() {
  //   const privateKeyStr = fs.readFileSync(path.join(__dirname, "./extdev_private_key"), "utf-8");
  const privateKeyStr = LOOM_PRIVATE_KEY;
  const privateKey = CryptoUtils.B64ToUint8Array(privateKeyStr);
  const publicKey = CryptoUtils.publicKeyFromPrivateKey(privateKey);
  const client = new Client(LOOM_NETWORK, LOOM_WRITE_URL, LOOM_READ_URL);
  client.txMiddleware = [new NonceTxMiddleware(publicKey, client), new SignedTxMiddleware(privateKey)];
  //   client.on("error", (msg) => {
  //     console.error("PlasmaChain connection error", msg);
  //   });

  return {
    account: LocalAddress.fromPublicKey(publicKey).toString(),
    web3js: new Web3(new LoomProvider(client, privateKey)),
    client,
  };
}

async function mapContracts({
  client,
  signer,
  tokenRinkebyAddress,
  tokenExtdevAddress,
  ownerExtdevAddress,
  rinkebyTxHash,
}) {
  const ownerExtdevAddr = Address.fromString(`${client.chainId}:${ownerExtdevAddress}`);
  const gatewayContract = await TransferGateway.createAsync(client, ownerExtdevAddr);
  const foreignContract = Address.fromString(`eth:${tokenRinkebyAddress}`);
  const localContract = Address.fromString(`${client.chainId}:${tokenExtdevAddress}`);

  const hash = soliditySha3(
    { type: "address", value: tokenRinkebyAddress.slice(2) },
    { type: "address", value: tokenExtdevAddress.slice(2) },
  );

  const foreignContractCreatorSig = await signer.signAsync(hash);
  const foreignContractCreatorTxHash = Buffer.from(rinkebyTxHash.slice(2), "hex");

  await gatewayContract.addContractMappingAsync({
    localContract,
    foreignContract,
    foreignContractCreatorSig,
    foreignContractCreatorTxHash,
  });
}

function loadRinkebyAccount() {
  // const privateKey = fs.readFileSync(path.join(__dirname, "./rinkeby_private_key"), "utf-8");
  const privateKey = ETHEREUM_PRIVATE_KEY;
  const web3js = new Web3(INFURA_PROVIDER_URL);
  const ownerAccount = web3js.eth.accounts.privateKeyToAccount(privateKey);
  web3js.eth.accounts.wallet.add(ownerAccount);
  return { account: ownerAccount, web3js };
}

const main = async () => {
  console.log("Mapping contracts...");
  const { web3js, account } = loadRinkebyAccount();
  const signer = new OfflineWeb3Signer(web3js, account);

  const { client, account: ownerExtdevAddress } = loadExtdevAccount();

  await mapContracts({
    client,
    signer,
    tokenRinkebyAddress: ethereumDaiAddress,
    tokenExtdevAddress: loomDaiAddress,
    ownerExtdevAddress,
    rinkebyTxHash: ethereumDaiDeployTxHash,
  });

  client.disconnect();
  console.log("Mapping contracts... done!");
};

main().catch(console.log);
