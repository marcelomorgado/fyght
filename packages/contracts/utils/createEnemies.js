/* eslint-disable no-console */
/* eslint-disable no-await-in-loop */
const Fyghters = artifacts.require("Fyghters");
const Layer2Dai = artifacts.require("Layer2Dai");

const APPROVAL_AMOUNT = `${100e18}`;
const DEPOSIT_AMOUNT = `${5e18}`;

// FIXME: This script stucks when some tx is rejected
module.exports = async (callback) => {
  const fyghters = await Fyghters.deployed();
  const dai = await Layer2Dai.deployed();

  const accounts = await web3.eth.getAccounts();
  const enemiesAccounts = accounts.slice(1, 5);

  const names = ["Bruce", "Chuck", "Jean", "Jackie"];

  console.log("Creating enemies...");

  for (let i = 0; i < enemiesAccounts.length; ++i) {
    const name = names[i];
    const from = enemiesAccounts[i];
    console.log(`Creating enemy... [name = ${name}, owner = ${from}]`);

    const { tx: tx1 } = await dai.mint(APPROVAL_AMOUNT, { from });
    console.log(`Tokens minted: ${tx1}`);

    const { tx: tx2 } = await dai.approve(fyghters.address, APPROVAL_AMOUNT, { from });
    console.log(`Fyghters contract allowed to spend tokens: ${tx2}`);

    const { tx: tx3 } = await fyghters.create(name, { from });
    await fyghters.deposit(i, DEPOSIT_AMOUNT, { from });
    console.log(`Fyghter created: ${tx3}`);
  }

  callback();
};
