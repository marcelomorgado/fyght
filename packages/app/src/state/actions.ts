import { StoreActionApi } from "react-sweet-state";
import { ethers, Event, ContractReceipt, BigNumber } from "ethers";
import { optimisticUpdate } from "./utils";
import { Skin, MIN_DEPOSIT } from "../constants";
const { getAddress } = ethers.utils;

type State = FyghtContext;
type StoreApi = StoreActionApi<State>;

export const fetchMyFyghter = () => async ({ setState, getState }: StoreApi): Promise<void> => {
  const {
    metamask: {
      contracts: { fyghters },
      account,
    },
  } = getState();

  const filter = fyghters.filters.FyghterCreated(getAddress(account), null, null);
  const logs = await fyghters.queryFilter(filter, 0, "latest");
  const [myFyghterId] = logs.map((l: Event) => l.args).map(({ id }: FyghterCreated) => id);

  if (myFyghterId) {
    const myFyghter = await fyghters.fyghters(myFyghterId);
    setState({ myFyghter });
  }
};

export const fetchAllEnemies = () => async ({ setState, getState }: StoreApi) => {
  const {
    metamask: {
      contracts: { fyghters },
      provider,
      account,
    },
    myFyghter,
  } = getState();

  if (!fyghters || !provider) {
    setState({ enemies: [] });
    return;
  }

  const loadEnemy = async (id: BigNumber): Promise<Enemy> => {
    const fyghter: Fyghter = await fyghters.fyghters(id);

    let winProbability = null;

    if (myFyghter && myFyghter.id) {
      const { id: myFyghterId } = myFyghter;
      winProbability = await fyghters.calculateWinProbability(myFyghterId, id);
    }
    return { fyghter, winProbability };
  };

  const filter = fyghters.filters.FyghterCreated(null, null, null);
  const logs = await fyghters.queryFilter(filter, 0, "latest");

  const enemiesIds = logs
    .map((l: Event) => l.args)
    .filter(({ owner }: FyghterCreated) => getAddress(owner) !== getAddress(account))
    .map(({ id }: FyghterCreated) => id);

  const enemiesPromises = enemiesIds.map((id: BigNumber) => loadEnemy(id));
  const enemies: Enemy[] = await Promise.all(enemiesPromises);

  setState({ enemies });
};

export const setErrorMessage = (errorMessage: string) => ({ setState, getState }: StoreApi): void => {
  const { messages } = getState();
  setState({ messages: { ...messages, errorMessage } });
};

export const setInfoMessage = (infoMessage: string) => ({ setState, getState }: StoreApi): void => {
  const { messages } = getState();
  setState({ messages: { ...messages, infoMessage } });
};

export const renameMyFyghter = (newName: string) => async ({ getState, setState }: StoreApi): Promise<void> => {
  const {
    myFyghter,
    metamask: {
      contracts: { fyghters },
    },
  } = getState();
  const { id: myFyghterId, name: oldName } = myFyghter;

  optimisticUpdate({
    doTransaction: () => fyghters.rename(myFyghterId, newName),
    onOptimistic: () => {
      setState({ myFyghter: { ...myFyghter, name: newName } });
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: { ...myFyghter, name: oldName } });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const createFyghter = (name: string) => async ({ getState, setState, dispatch }: StoreApi): Promise<void> => {
  const {
    metamask: {
      contracts: { fyghters, dai },
    },
  } = getState();

  optimisticUpdate({
    doTransaction: async () => {
      // TODO: Create a button for this? (Refs: https://github.com/marcelomorgado/fyght/issues/118)
      await dai.mint(MIN_DEPOSIT);
      // TODO: Call for approval only if needed (Refs: https://github.com/marcelomorgado/fyght/issues/118)
      await dai.approve(fyghters.address, MIN_DEPOSIT);
      return fyghters.create(name);
    },
    onOptimistic: () => {
      const myFyghter: Fyghter = {
        id: null,
        skin: Skin.NAKED,
        name,
        xp: BigNumber.from("1"),
        balance: BigNumber.from(MIN_DEPOSIT),
      };
      setState({ myFyghter });
    },
    onSuccess: async () => {
      dispatch(fetchMyFyghter());
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: null });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const changeMyFyghterSkin = (newSkin: string) => async ({ getState, setState }: StoreApi): Promise<void> => {
  const {
    myFyghter,
    metamask: {
      contracts: { fyghters },
    },
  } = getState();
  const { id: myFyghterId, skin: oldSkin } = myFyghter;

  optimisticUpdate({
    doTransaction: () => fyghters.changeSkin(myFyghterId, newSkin),
    onOptimistic: () => {
      setState({ myFyghter: { ...myFyghter, skin: newSkin } });
    },
    onError: (errorMessage: string) => {
      setState({ myFyghter: { ...myFyghter, skin: oldSkin } });
      setErrorMessage(errorMessage);
    },
    getState,
  });
};

export const challengeAnEnemy = (enemyId: BigNumber, whenFinish: () => void) => async ({
  getState,
}: StoreApi): Promise<void> => {
  const {
    myFyghter: { id: myFyghterId },
    metamask: {
      contracts: { fyghters },
    },
  } = getState();

  optimisticUpdate({
    doTransaction: () => fyghters.challenge(myFyghterId, enemyId),
    onSuccess: (receipt: ContractReceipt) => {
      const [log] = receipt.logs
        .map((log: Event) => fyghters.interface.parseLog(log))
        .filter(({ name }) => name == "ChallengeOccurred")
        .map(({ args }) => args);
      const [myFyghterId, , winnerId] = log;
      if (winnerId.eq(myFyghterId)) {
        setInfoMessage("You won!");
      } else {
        setInfoMessage("You lose!");
      }

      fetchMyFyghter();
      // TODO: fetch only the enemy
      fetchAllEnemies();
      whenFinish();
    },
    onError: (errorMessage: string) => {
      setErrorMessage(errorMessage);
      whenFinish();
    },
    getState,
  });
};

const setMetamaskNetworkId = (networkId: number) => ({ setState, getState }: StoreApi) => {
  const { metamask } = getState();
  setState({ metamask: { ...metamask, networkId } });
};

export const setMetamaskAccount = (account: string) => ({ setState, getState }: StoreApi) => {
  const { metamask } = getState();
  setState({ metamask: { ...metamask, account } });
};

export const initializeMetamask = () => async ({ setState, getState, dispatch }: StoreApi) => {
  const { metamask } = getState();

  const { ethereum } = window;
  if (ethereum) {
    ethereum.autoRefreshOnNetworkChange = false;
  }
  // Note: Parcel doesn't support process.env es6 destructuring
  //

  // Refs:
  // https://github.com/parcel-bundler/parcel/issues/2299#issuecomment-439768971
  // https://en.parceljs.org/env.html

  // eslint-disable-next-line no-undef
  const FYGHTERS_CONTRACT_ADDRESS = process.env.FYGHTERS_CONTRACT_ADDRESS;
  // eslint-disable-next-line no-undef
  const DAI_CONTRACT_ADDRESS = process.env.DAI_CONTRACT_ADDRESS;

  if (ethereum) {
    // Note: The metamask docs recommends to use the 'chainChanged' event instead but it isn't working
    // See more: https://docs.metamask.io/guide/ethereum-provider.html#methods-new-api
    ethereum.on("networkChanged", (networkId: number) => {
      dispatch(setMetamaskNetworkId(networkId));
    });

    ethereum.on("accountsChanged", ([account]: string[]) => {
      dispatch(setMetamaskAccount(account));
    });

    const provider = new ethers.providers.Web3Provider(ethereum);
    const signer = provider.getSigner();
    const FYGHTERS_CONTRACT_ABI = require("../contracts/Fyghters.json").abi;
    const DAI_CONTRACT_ABI = require("../contracts/Dai.json").abi;

    const fyghters = new ethers.Contract(FYGHTERS_CONTRACT_ADDRESS, FYGHTERS_CONTRACT_ABI, signer);

    const dai = new ethers.Contract(DAI_CONTRACT_ADDRESS, DAI_CONTRACT_ABI, signer);

    const [account] = await ethereum.enable();

    // TODO: It isn't working sometimes
    const { networkVersion: networkId } = ethereum;
    setState({
      metamask: {
        ...metamask,
        contracts: { fyghters, dai },
        ethereum,
        account,
        provider,
        networkId,
        loading: false,
      },
    });
  }
};
