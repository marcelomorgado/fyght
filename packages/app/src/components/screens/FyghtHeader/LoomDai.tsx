import React from "react";
import { Col, Button, Typography } from "antd";
import { useFyghtState } from "../../../state";
import { formatDai } from "../../../helpers";
import { DEPOSIT_TO_LOOM_AMOUNT } from "../../../constants";

const { Text } = Typography;

export const LoomDai: React.FC = () => {
  const [
    {
      daiBalances: {
        loomBalance: { loading, amount: balanceInWei },
      },
    },
    { depositToLoom, withdrawFromLoom },
  ] = useFyghtState();

  return (
    <>
      <Col span={3}>
        <Text code style={{ color: "#fff" }}>
          {`FyGHT Balance ${formatDai(balanceInWei)}`}
        </Text>
      </Col>
      <Col span={2}>
        <Button type="primary" onClick={depositToLoom} loading={loading}>
          {`Deposit ${formatDai(DEPOSIT_TO_LOOM_AMOUNT)}`}
        </Button>
      </Col>
      <Col span={2}>
        <Button type="primary" onClick={withdrawFromLoom} loading={loading}>
          Withdraw All
        </Button>
      </Col>
    </>
  );
};
