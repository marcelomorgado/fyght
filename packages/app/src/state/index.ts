import { createStore, createHook } from "react-sweet-state";
import * as actions from "./actions";
import { BigNumber } from "ethers/utils";

type State = FyghtState;
type Actions = typeof actions;

const initialState: State = {
  myFyghter: null,
  enemies: [],
  metamask: {
    networkId: null,
    ethereumAccount: null,
    loomAccount: null,
    ethereum: null,
    contracts: { fyghters: null, loomDai: null, ethereumDai: null, ethereumGateway: null, loomGateway: null },
    loomProvider: null,
    loomClient: null,
    ethereumProvider: null,
    loading: true,
  },
  daiBalances: {
    ethereumBalance: { amount: new BigNumber(0), loading: false },
    loomBalance: { amount: new BigNumber(0), loading: false },
  },
};

const Store = createStore<State, Actions>({
  initialState,
  actions,
});

export const useFyghtState = createHook(Store);
