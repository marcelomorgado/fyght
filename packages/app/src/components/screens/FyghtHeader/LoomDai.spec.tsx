import * as React from "react";
import { shallow } from "enzyme";
import * as FyghtState from "../../../state";
import { LoomDai } from "./LoomDai";
import { BigNumber } from "ethers/utils";

describe("LoomDai", () => {
  test("should render the component", () => {
    jest
      .spyOn(FyghtState, "useFyghtState")
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .mockImplementation((): any => [
        {
          daiBalances: {
            loomBalance: { amount: new BigNumber(10), loading: false },
          },
        },
        { depositToLoom: jest.fn(), withdrawFromLoom: jest.fn() },
      ]);

    expect(shallow(<LoomDai />)).toMatchSnapshot();
  });
});
