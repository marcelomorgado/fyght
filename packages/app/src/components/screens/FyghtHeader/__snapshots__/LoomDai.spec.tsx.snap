// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoomDai should render the component 1`] = `
<Space>
  <Text
    code={true}
    style={
      Object {
        "color": "#fff",
      }
    }
  >
    FyGHT Balance $NaN
  </Text>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[MockFunction]}
    type="primary"
  >
    Deposit $10.00
  </Button>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[MockFunction]}
    type="primary"
  >
    Withdraw All
  </Button>
</Space>
`;
